pkg/koordlet/statesinformer/states_gpu.go

```diff
@@ -0,0 +1,207 @@
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	if nodeResource.Error != nil {
+		klog.Errorf("failed to get node resource metric, err: %v", nodeResource.Error)
+		return
+	}
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.V(5).Info("no gpu device found")
+		return
+	}
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
+	for _, gpu := range nodeResource.Metric.GPUs {
+		health := true
+		s.gpuMutex.RLock()
+		if _, ok := s.unhealthyGPU[gpu.DeviceUUID]; ok {
+			health = false
+		}
+		s.gpuMutex.RUnlock()
+		deviceInfos = append(deviceInfos, schedulingv1alpha1.DeviceInfo{
+			UUID:   gpu.DeviceUUID,
+			Minor:  gpu.Minor,
+			Type:   schedulingv1alpha1.GPU,
+			Health: health,
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.DecimalSI),
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+		})
+	}
+	blocker := true
+	device := &schedulingv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: copyNode.Name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "v1",
+					Kind:               "Node",
+					Name:               copyNode.Name,
+					UID:                copyNode.UID,
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		Spec: schedulingv1alpha1.DeviceSpec{
+			Devices: deviceInfos,
+		},
+	}
+	// TODO: compare diff before update
+	_, err := s.deviceClient.Update(context.TODO(), device, metav1.UpdateOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to update device %s, err: %v", copyNode.Name, err)
+			return
+		}
+		_, err = s.deviceClient.Create(context.TODO(), device, metav1.CreateOptions{})
+		if err != nil {
+			klog.Errorf("failed to create device %s, err: %v", copyNode.Name, err)
+			return
+		}
+	}
+}
+
+func (s *statesInformer) GPUHealCheck(stopCh <-chan struct{}) {
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		klog.Errorf("unable to get device count: %v", nvml.ErrorString(ret))
+		return
+	}
+	if count == 0 {
+		klog.Errorf("no gpu device found")
+		return
+	}
+	devices := []string{}
+	for deviceIndex := 0; deviceIndex < count; deviceIndex++ {
+		gpudevice, ret := nvml.DeviceGetHandleByIndex(deviceIndex)
+		if ret != nvml.SUCCESS {
+			klog.Errorf("unable to get device at index %d: %v", deviceIndex, nvml.ErrorString(ret))
+			continue
+		}
+		uuid, ret := gpudevice.GetUUID()
+		if ret != nvml.SUCCESS {
+			klog.Errorf("failed to get device uuid at index %d, err: %v", deviceIndex, nvml.ErrorString(ret))
+		}
+		devices = append(devices, uuid)
+	}
+	unhealthyChan := make(chan string)
+	go checkHealth(stopCh, devices, unhealthyChan)
+	klog.Info("start to do gpu health check")
+	for d := range unhealthyChan {
+		// FIXME: there is no way to recover from the Unhealthy state.
+		s.gpuMutex.Lock()
+		s.unhealthyGPU[d] = struct{}{}
```

Comment:
I don't find move gpu minor out of unhealthyGPU logic if card turns good.

Reply:
from the original code of asi device plugin，there's a comment

// FIXME: there is no way to recover from the Unhealthy state.

please fixme if error

Reply:
https://github.com/NVIDIA/k8s-device-plugin/blob/master/internal/rm/health.go can't change unhealthy to healthy too... so just follow it.